### PR TITLE
Add experimental vLLM acceleration demo

### PR DESCRIPTION
**Pull request by dibbla**
## Current problem
The speed of current implementation can be improved by utilizing inference engines.

## Solution
Implement an experiment demo of using vLLM to accelerate the algorithm. While the general design is not yet fully implemented in PR, we provide a viable version to accelerate the research community.

### Walkthrough

This implementation fine-tunes LLMs on the Countdown task using ES with vLLM for high-throughput generation and Ray for multi-GPU orchestration.

- Architecture
  - One vLLM engine per GPU using Ray placement groups for isolation
  - Engines are ES-wrapped vLLM actors with `distributed_executor_backend="ray"` and `utils.worker_extn.WorkerExtension`.
  - Base model is loaded once via Hugging Face, saved to disk, then shared by all engines

- Data and Generation
  - Loads `countdown/data/countdown.json`
  - Deterministic sampling

- ES Loop (per iteration)
  1. Snapshot engine baseline weights
  2. Evaluate a population by applying Gaussian perturbations (`perturb_self_weights(seed, sigma, ...)`) and round-robin scheduling across engines.
  3. Score generations with `reward_function`, normalize rewards, and log to TensorBoard.
  4. Restore engines to baseline (`restore_self_initial_weights`), then apply weighted updates across seeds using normalized rewards.
  5. Re-snapshot the new baseline for the next iteration.

- Persistence and Cleanup
  - Logs under `--experiment_dir`, TensorBoard supported.
  - Final synchronized weights saved from engine 0 to `model_saves/final_model_iteration_<N>/pytorch_model.pth`.
  - Graceful shutdown on SIGINT/SIGTERM, Ray placement groups cleaned up.

- Key CLI Flags
  - `--model_name`, `--sigma` (exploration), `--alpha` (LR), `--population_size`,
    `--num_engines` (GPUs), `--num_iterations`, `--experiment_dir`, `--cuda_devices`.
  - `--cuda_devices` sets process-level visibility; engines clear it so Ray assigns GPUs.

### CODE REVIEW

File: es_fine-tuning_countdown_accl.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+def evaluate_countdown_handle(llm, task_datas):
+    """Return a generation handle so we can schedule round-robin."""
+    start = time.time()
+    prompts = [d["context"] for d in task_datas]
+    sampling_params = SamplingParams(
+        temperature=0.0,
+        seed=42,
+        max_tokens=1024,
+    )
+    handle = llm.generate.remote(prompts, sampling_params, use_tqdm=False)
```
[the remark this answers is not in the document]
Comment (dibbla) [reply es_fine-tuning_countdown_accl.py]: Here we do not have the option to pass attention masks ([ref](https://github.com/vllm-project/vllm/issues/17869)), and the batching / scheduling is handled by vLLM.

The `llm.generate` here is different from HF's `transformers.GenerationMixin.generate()`, instead it is the generation method from vLLM's `LLM` class ([vLLM doc](https://docs.vllm.ai/en/latest/examples/offline_inference/basic.html)).

[the remark this answers is not in the document]
Comment (conorfhayes) [reply es_fine-tuning_countdown_accl.py]: Yeah, the model should be trained with padding_side="left". We are then attending to those tokens when doing inference. Does that matter? This will also happen for RL implementations with vLLM. I suppose we can compare performance with and without vLLM for our ES. This is something worth exploring, but again maybe the speedup is worth the potential cost in performance.

File: es_fine-tuning_countdown_accl.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        # Log generation statistics
+        writer.add_scalar("reward/mean", mean_reward, i)
+        writer.add_scalar("reward/std", std_reward, i)
+        writer.add_scalar("reward/min", min_reward, i)
+        writer.add_scalar("reward/max", max_reward, i)
+
+        # Bring all engines to the exact same baseline
+        ray.get(
+            [
+                llm.collective_rpc.remote("restore_self_initial_weights")
```
[the remark this answers is not in the document]
Comment (dibbla) [reply es_fine-tuning_countdown_accl.py]: Now we broadcast from engine 0 instead of direct save/load in [exp: synchronize weights by broadcast](https://github.com/VsonicV/es-fine-tuning-paper/pull/17/commits/19984838bfb12a045c2efe0aa7446500aa0ba0ed)

File: es_fine-tuning_countdown_accl.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+                    llm.collective_rpc.remote(
+                        "perturb_self_weights", args=(seed, 1.0, coeff, False)
+                    )
+                    for llm in engines
+                ]
+            )
+
+        # Re-snapshot the new baseline for the next iteration
+        ray.get(
+            [llm.collective_rpc.remote("save_self_initial_weights") for llm in engines]
```
[the remark this answers is not in the document]
Comment (dibbla) [reply es_fine-tuning_countdown_accl.py]: Resolved

File: utils/worker_extn.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    def perturb_self_weights(self, seed, sigma_or_scale, coeff=1.0, negate=False):
+        """
+        Add noise(seed) scaled by sigma_or_scale * coeff (or subtract when negate=True).
+        - For exploration:  perturb_self_weights(seed, SIGMA, 1.0, False)
+          and restore with restore_self_weights(seed, SIGMA) as before.
+        - For ES update:   perturb_self_weights(seed, 1.0, coeff, False)
+          where coeff = ALPHA/POPULATION_SIZE * norm_reward.
+        """
+        scale = float(sigma_or_scale) * float(coeff)
```
[the remark this answers is not in the document]
Comment (dibbla) [reply utils/worker_extn.py]: 2 params now merged.

**Review by VsonicV — APPROVED** [review]
lgtm
